### PR TITLE
Serverless function creation with Pipelines

### PR DESCRIPTION
**Pull request by vikram-raj**
Story: https://issues.redhat.com/browse/ODC-7235

Screenshots/GIF:
- Create Serverless function via Import from Git with Pipeline
Create-serverless-function-via-Import-from-Git-with-Pipeline

- Create Serverless function via Create Serverless form with Pipeline
Create-serverless-function-via-Create-Serverless-form-with-Pipeline gif

Test setup:
- Install Pipeline and Serverless Operator and create CRs
- Add all ClusterTasks and Pipeline from this gist https://gist.github.com/vikram-raj/386b4eca5193036b0ca29471f6e513a5
- Navigate to Import from Git form and enter Git URL https://github.com/vikram-raj/hello-func-node
- Check the Pipeline checkbox and click on create
- Check in the Topology that the Serverless function is get created and a Pipeline decorator is also attached to it.
- Also the Create Serverless function form by entering Git URL https://github.com/vikram-raj/hello-func-node and enable Pipeline in the form

### CODE REVIEW

**Comment by nainaz** [comment]
This looks good. Thank you @serenamarie125 and @vikram-raj 
@lance we need to make sure we are providing tasks through tekton hub or through serverless opertor. 
@zroubalik

**Review by Lucifergene — COMMENTED** [review]
@vikram-raj The Git Import Flow looks good from my side.
But, found a small bug in the Create Serverless function form. 

The issue is regarding the **Add Pipeline** Checkbox. When there are 2 pipelines displayed in the dropdown menu, selecting one, unchecks the Add Pipeline checkbox.

Attached is a video for reference:

https://user-images.githubusercontent.com/47265560/217908221-106645dc-cda1-40ef-b974-762ce097539e.mp4

**Note:** I used the Git Import flow first and then tested the Create Serverless function form.

**Comment by vikram-raj** [comment]
> @vikram-raj The Git Import Flow looks good from my side. But, found a small bug in the Create Serverless function form.
> 
> The issue is regarding the **Add Pipeline** Checkbox. When there are 2 pipelines displayed in the dropdown menu, selecting one, unchecks the Add Pipeline checkbox.
> 
> Attached is a video for reference:
> 
>  Tab-Create.Serverless.function.Red.Hat.OpenShift.mp4 
> **Note:** I used the Git Import flow first and then tested the Create Serverless function form.

@Lucifergene Good catch. This is in our current master in Import from Git form. Can you please raise this Bug and we will fix it in another PR. 
Peek 2023-02-10 19-10

**Comment by debsmita1** [comment]
tested this locally
/lgtm

**Comment by debsmita1** [comment]
@vikram-raj can you add an e2e test for "Create serverless function using Create Serverless function form without Pipeline" ?

**Comment by debsmita1** [comment]
@vikram-raj can you add an e2e test for "Create serverless function using Create Serverless function form without Pipeline" ?

**Comment by vikram-raj** [comment]
@debsmita1  Added e2e tests. PTAL again. Thanks

**Comment by vikram-raj** [comment]
Propagating docs and px acks from the Epic
/label docs-approved
/label px-approved

File: frontend/packages/knative-plugin/integration-tests/support/step-definitions/serverless/serverless-function.ts [hunk 1]
```diff
@@ -44,3 +45,15 @@ Then('user switches to the {string} tab', (tab: string) => {
...
   cy.byLegacyTestID('submit-button').click();
 });
+
+When('user selects Add Pipeline checkbox in Pipelines section', () => {
+  gitPage.selectAddPipeline();
+});
+
+Then('user is able to see PipelineRuns', () => {
+  cy.byTestID('pipeline-overview').should('be.visible');
+});
```
Comment (sanketpathak) [inline frontend/packages/knative-plugin/integration-tests/support/step-definitions/serverless/serverless-function.ts:55]:
```suggestion
Then('user is able to see PipelineRuns in the {string} tab', (tab: string) => {
   topologySidePane.selectTab(tab)
  cy.byTestID('pipeline-overview').should('be.visible');
});
```

File: frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature [hunk 2]
```diff
@@ -77,3 +75,60 @@ Feature: Creation and Visualisation of serverless fuctions
...
+        @regression @odc-6360
+        Scenario Outline: Create Serverless Function from the Import from Git Form on Add page with Pipeline: SF-01-TC07
+            Given user is at Add page
+              And user is at Import from Git form
+             When user enters Git Repo URL as "<git_url>"
+              And user enters Name as "<workload_name>"
+              And user selects Add Pipeline checkbox in Pipelines section
+              And user clicks Create button on Add page
+             Then user will be redirected to Topology page
+              And user clicks on Topology list view
```
Comment (sanketpathak) [inline frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature:88]:
```suggestion
             And user clicks on List view button
```

File: frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature [hunk 3]
```diff
@@ -77,3 +75,60 @@ Feature: Creation and Visualisation of serverless fuctions
...
+              And user is able to see Type as Function
+              And user switches to the "Resources" tab
+              And user is able to see PipelineRuns
+
+        Examples:
+                  | git_url                                       | workload_name                        |
+                  | https://github.com/vikram-raj/hello-func-node | import-form-hello-func-node-pipeline |
+
+        @regression @odc-6360
+        Scenario Outline: Create serverless function using Create Serverless function form with Pipeline: SF-01-TC08
```
Comment (sanketpathak) [inline frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature:101]: You can edit this as well based on the above comments

File: frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature [hunk 4]
```diff
@@ -77,3 +75,60 @@ Feature: Creation and Visualisation of serverless fuctions
...
+                  | git_url                                       | workload_name                                            |
+                  | https://github.com/vikram-raj/hello-func-node | create-serverless-function-form-hello-func-node-pipeline |
+
+        @regression @odc-6360
+        Scenario Outline: Pipeline section should not present in Create Serverless function form if Pipeline is not available: SF-01-TC09
+            Given user is at Add page
+             When user clicks on Create Serverless function card
+              And user enters git url "<git_url>"
+              And user is able to see builder image version dropdown
+              And user is able to see the runtime details
```
Comment (sanketpathak) [inline frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature:128]: These two steps are not needed and can be removed

**Comment by vikram-raj** [comment]
@sanketpathak updated the test scenarios as per your comment. PTAL again. Thanks

File: frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature [hunk 5]
```diff
@@ -77,3 +75,55 @@ Feature: Creation and Visualisation of serverless fuctions
         Examples:
                   | git_url                                           | workload_name       |
                   | https://github.com/vikram-raj/hello-func-node-env | hello-func-node-env |
+
+        @regression @odc-6360
+        Scenario Outline: Create Serverless Function from the Import from Git Form on Add page with Pipeline: SF-01-TC07
+            Given user is at Add page
```
Comment (sanketpathak) [inline frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature:81]:
```suggestion
            Given user has installed Openshift Pipelines Operator
              And user is at Add page
```

File: frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature [hunk 6]
```diff
@@ -77,3 +75,55 @@ Feature: Creation and Visualisation of serverless fuctions
...
+              And user is able to see Type as Function
+              And user is able to see PipelineRuns in the "Resources" tab
+
+        Examples:
+                  | git_url                                       | workload_name                        |
+                  | https://github.com/vikram-raj/hello-func-node | import-form-hello-func-node-pipeline |
+
+        @regression @odc-6360
+        Scenario Outline: Create serverless function using Create Serverless function form with Pipeline: SF-01-TC08
+            Given user is at Add page
```
Comment (sanketpathak) [inline frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature:100]:
```suggestion
            Given user has installed Openshift Pipelines Operator
              And user is at Add page
```

File: frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature [hunk 7]
```diff
@@ -77,3 +75,58 @@ Feature: Creation and Visualisation of serverless fuctions
         Examples:
                   | git_url                                           | workload_name       |
                   | https://github.com/vikram-raj/hello-func-node-env | hello-func-node-env |
+
+        @regression @odc-6360
+        Scenario Outline: Create Serverless Function from the Import from Git Form on Add page with Pipeline: SF-01-TC07
+            Given user has installed OpenShift Pipelines Operator
```
Comment (sanketpathak) [inline frontend/packages/knative-plugin/integration-tests/features/serverless/serverless-function.feature:81]:
```suggestion
// Below cluster tasks and pipeline yaml implementation in given statement to be removed when Serverless new version <version_number> is released
            Given user has created ClusterTasks "func-buildpacks" from yaml "test-data/func-buildpacks.yaml"
              And user has created ClusterTasks "func-deploy" from yaml "test-data/func-deploy.yaml"
              And user has created ClusterTasks "func-s2i" from yaml "test-data/func-s2i.yaml"
              And user has created ClusterTasks "viewer-s2i-git-pipeline" from yaml "test-data/viewer-s2i-git-pipeline.yaml"
              And user has installed OpenShift Pipelines Operator
```

**Comment by sanketpathak** [comment]
Verified changes with kube:admin and developer role on cluster created using pr
[Screencast from 02-17-2023 08:13:24 PM.webm](https://user-images.githubusercontent.com/10252230/220286449-6de77b1f-90d6-4c5f-b216-81700c4c08e6.webm)
/label qe-approved

**Pull request merged**